### PR TITLE
[WIP] Add secio & move to pull-streams

### LINKED ISSUE #63: Add secio
I'm currently working on this, work can be tracked here

- https://github.com/ipfs/js-libp2p-crypto
- https://github.com/ipfs/js-libp2p-secio

### PR DESCRIPTION
**Pull request by dignifiedquire**
Not finished yet

Closes #63

### CODE REVIEW

**Comment by dignifiedquire** [comment]
I'm pretty close to passing tests, there is just one issue left to fix. When initiating the `secio` protocol multistream-select seems to loose one message, which is very unfortunate. 

``` sh
-->  /multistream/1.0.0
<--  /multistream/1.0.0
<--  /secio/1.0.0
listener ready
handle: QmVg942ZGNuGw7AFpiQAPsdRxzCK4J9e15NZGEBaixowQV
  libp2p:secio 1. propose - start +0ms
  libp2p:secio 1. propse - writing proposal +1ms
writing:  <Buffer f3 02 0a 10 98 a4 3a fb f0 07 e4 74 09 8c 9c 14 28 43 b9 6a 12 ab 02 08 00 12 a6 02 30 82 01 22 30 0d 06 09 2a 86 48 86 f7 0d 01 01 01 05 00 03 82 01 ... >
  libp2p:secio 1. propose - proposal size 373 +8ms
I shall read: 373
-->  /secio/1.0.0
dial: QmRRTGZEykibGGAEpSDh3fK2UnL746x5zHeQzNzERc74uM
  libp2p:secio 1. propose - start +4ms
  libp2p:secio 1. propse - writing proposal +0ms
writing:  <Buffer f3 02 0a 10 1b c0 15 68 22 cd c6 e9 0d 99 7e ae 4e 64 d8 f3 12 ab 02 08 00 12 a6 02 30 82 01 22 30 0d 06 09 2a 86 48 86 f7 0d 01 01 01 05 00 03 82 01 ... >
  libp2p:secio 1. propose - proposal size 373 +6ms
I shall read: 373
reading <Buffer 0d 2f 73 65 63 69 6f 2f 31 2e 30 2e 30 0a f3 02 0a 10 98 a4 3a fb f0 07 e4 74 09 8c 9c 14 28 43 b9 6a 12 ab 02 08 00 12 a6 02 30 82 01 22 30 0d 06 09 ... >
reading msg <Buffer f3 02 0a 10 1b c0 15 68 22 cd c6 e9 0d 99 7e ae 4e 64 d8 f3 12 ab 02 08 00 12 a6 02 30 82 01 22 30 0d 06 09 2a 86 48 86 f7 0d 01 01 01 05 00 03 82 01 ... >
  libp2p:secio 1. propse - reading proposal +4ms <Buffer 0a 10 1b c0 15 68 22 cd c6 e9 0d 99 7e ae 4e 64 d8 f3 12 ab 02 08 00 12 a6 02 30 82 01 22 30 0d 06 09 2a 86 48 86 f7 0d 01 01 01 05 00 03 82 01 0f 00 ... >
  libp2p:secio 1.1 identify +1ms
  libp2p:secio 1.1 identify - QmVg942ZGNuGw7AFpiQAPsdRxzCK4J9e15NZGEBaixowQV - identified remote peer as QmRRTGZEykibGGAEpSDh3fK2UnL746x5zHeQzNzERc74uM +6ms
  libp2p:secio 1.2 selection +0ms
  libp2p:secio 1. propose - finish +1ms
  libp2p:secio 2. exchange - start +1ms
  libp2p:secio 2. exchange - writing exchange +0ms
writing:  <Buffer c6 02 0a 41 04 87 54 67 c3 15 ba f1 b6 ff b9 c3 0c 6f 8a e4 16 fc b6 7a 54 51 20 e6 ee 5d 63 c1 bd 89 f2 b0 59 15 4b dd 1b 4e 50 fc c2 81 24 42 79 96 ... >
I shall read: 328

# hangs..
```

The message that is lost is the first one with `writing..`

**Comment by dignifiedquire** [comment]
Adding the below snippet to https://github.com/diasdavid/js-multistream/blob/master/src/listener.js#L62

``` js
      if (handlers[msg]) {
        // Protocol supported, ACK back
        encode.write(new Buffer(msg + '\n'));

        if (conn.cork) {
          conn.cork()

          setTimeout(() => {
            conn.uncork()
          }, 100)
        }
        console.log('listener ready')
        return handlers[msg](conn);
```

fixes the issue.

**Comment by dignifiedquire** [comment]
Depends on
- https://github.com/libp2p/js-libp2p-secio/pull/8
- https://github.com/dignifiedquire/pull-length-prefixed/pull/3
- https://github.com/multiformats/js-multistream/pull/19
- https://github.com/libp2p/js-libp2p-spdy/pull/16
- https://github.com/libp2p/js-libp2p-tcp/pull/20
- https://github.com/libp2p/js-libp2p-identify/pull/2

**Comment by daviddias** [comment]
@dignifiedquire just to confirm, this just depends on merges and releases, all the rest is 👌 ?

**Comment by daviddias** [comment]
What was the magic to not need `node-forge` anymore? `libp2p-crypto`? It doesn't require to be injected every single time anymore? @haadcode will be happy :)

File: src/connection.js [hunk 1]
```diff
@@ -35,7 +38,7 @@ module.exports = function connection (swarm) {
               ms.select(identify.multicodec, (err, conn) => {
                 if (err) { return cb(err) }
 
-                identify.exec(conn, (err, peerInfo, observedAddrs) => {
+                identify.listen(conn, (err, peerInfo, observedAddrs) => {
```
Comment (daviddias) [inline src/connection.js]: note: This will change to 'dialer', once identify CR is applied

File: src/dial.js [hunk 2]
```diff
@@ -97,13 +101,24 @@ module.exports = function dial (swarm) {
...
             ms.handle(conn, (err) => {
               if (err) {
                 return cb(err)
               }
-              ms.select('/plaintext/1.0.0', cb)
+
+              const id = swarm._peerInfo.id
+              if (id.privKey == null || swarm.encrypt === false) {
+                return ms.select(tags.plaintex, cb)
+              }
```
Comment (daviddias) [inline src/dial.js]: This should be a Swarm option. In the same fashion that we do .connection.reuse(), we should do .connection.crypto(secio) and then check if it has the ingredients necessary.

Comment (daviddias) [inline src/dial.js]: and also, .connection.plaintext(false) to disable plaintext mode

**Comment by dignifiedquire** [comment]
> @dignifiedquire just to confirm, this just depends on merges and releases, all the rest is 👌 ?

Secio yes, but some of the other tests do not yet pass with all the new pull-stream things. Especially webrtc as that isn't migrated yet, and I'm having some issues with websockets, probably due to the halfOpen issue.

File: src/dial.js (hunk 2 above)
Comment (dignifiedquire) [inline src/dial.js]: Why should this be a method? It's just a matter of setting `swarm.encrypt  = true` right now to control this.

**Comment by dignifiedquire** [comment]
> What was the magic to not need node-forge anymore? libp2p-crypto? It doesn't require to be injected every single time anymore? @haadcode will be happy :)

This: https://github.com/libp2p/js-libp2p-swarm/blob/56e558ea86c1eb5b5355ade8cda9652d1589c408/.aegir.js#L9-L12

**Comment by daviddias** [comment]
> This: https://github.com/libp2p/js-libp2p-swarm/blob/56e558ea86c1eb5b5355ade8cda9652d1589c408/.aegir.js#L9-L12

Almost @haadcode, almost :) :D

**Comment by dignifiedquire** [comment]
All tests are now ported and mostly passing. Only some related to websockets are still failing:
- https://github.com/libp2p/js-libp2p-swarm/blob/ea9c93c48e08f4d9c6f6bd53e1721f9f688aef26/test/09-swarm-with-muxing.node.js#L165
- https://github.com/libp2p/js-libp2p-swarm/blob/ea9c93c48e08f4d9c6f6bd53e1721f9f688aef26/test/browser-02-swarm-with-muxing-plus-websockets.js#L49

**Comment by dignifiedquire** [comment]
Started doing some fixes for the websocket problems in 
- https://github.com/pull-stream/pull-ws/pull/14 and
- https://github.com/libp2p/js-libp2p-websockets/pull/18/commits/d6c6aeaf196c8399abc5f7e00549eeace94d6382

File: src/dial.js (hunk 2 above)
Comment (daviddias) [reply src/dial.js]: so that later we can use the method to pick different crypto channels

**Comment by daviddias** [comment]
~~@dignifiedquire having an hard time getting the stream muxing tests to pass:~~ **nvm, got it to pass, it was the circular npm linking**

```
    stream muxing with spdy (on TCP)
      ✓ add
      3) handle + dial on protocol
      4) dial to warm conn
      5) dial on protocol, reuse warmed conn
      6) enable identify to reuse incomming muxed conn
      7) with Identify, do getPeerInfo
      - make sure it does not blow up when the socket is closed
      - blow up a socket, with WebSockets
      ✓ close one end, make sure the other does not blow (1006ms)

```

**Comment by daviddias** [comment]
Last things before merge
- [x] point to lib instead of source
- [x] add documentation about pull-streams (same as libp2p-tcp)
- [x] make the encrypted channel be a option set by a func just like connection.reuse - https://github.com/libp2p/js-libp2p-swarm/pull/67#discussion_r77364562
- [x] With https://github.com/libp2p/js-libp2p-identify/pull/2, listener and dialer need to be switched (as they should be)

File: src/dial.js [hunk 3]
```diff
@@ -97,13 +101,24 @@ module.exports = function dial (swarm) {
...
             const ms = new multistream.Dialer()
             ms.handle(conn, (err) => {
               if (err) {
                 return cb(err)
               }
-              ms.select('/plaintext/1.0.0', cb)
+
+              const id = swarm._peerInfo.id
+              if (id.privKey == null || swarm.encrypt === false) {
+                return ms.select(tags.plaintex, cb)
```
Comment (daviddias) [inline src/dial.js]: this typo is what was causing my issue in libp2p-ipfs

**Comment by dignifiedquire** [comment]
> make the encrypted channel be a option set by a func just like connection.reuse

This is done, needs the new secio api: https://github.com/libp2p/js-libp2p-secio/pull/17

**Comment by dignifiedquire** [comment]
All tests pass locally when using the above mentioned secio branch plus https://github.com/libp2p/js-libp2p-webrtc-star/pull/19

**Comment by daviddias** [comment]
@dignifiedquire why do we have coveralls badges all over again?

**Comment by dignifiedquire** [comment]
> @dignifiedquire why do we have coveralls badges all over again?

trying to disable them one at a time, but they are enabled by default :(

**Pull request merged**